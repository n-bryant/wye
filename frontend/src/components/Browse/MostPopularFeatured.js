import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";
import { withWidth } from "@material-ui/core/";

import Grid from "@material-ui/core/Grid";

import PaginationWidget from "../PaginationWidget";
import FeaturedWidget from "./FeaturedWidget";
import BrowseCard from "./BrowseCard";
import styles from "./MostPopularFeatured.styles";

export const MOST_POPULAR_TITLE = "Trending";
export const MOST_POPULAR_SUBTITLE = "Most popular by recent playtime";

/**
 * renders a FeaturedWidget with cards representing the 5 most popular multiplayer games
 */
export const MostPopularFeatured = props => {
  const classnames = MostPopularFeatured.classnames(props);
  const { items, width } = props;
  const [currentPage, setCurrentPage] = React.useState(1);

  const featuredItem = items[0];
  const subFeaturedItems = items.slice(1);
  return (
    <FeaturedWidget
      className={classnames.root()}
      title={MOST_POPULAR_TITLE}
      subTitle={MOST_POPULAR_SUBTITLE}
    >
      {!["xs", "sm"].some(val => val === width) && items.length > 1 ? (
        <Grid
          className={classnames.element("featuredItemContainer")}
          container
          justify={"space-between"}
          alignItems={"center"}
        >
          <Grid
            className={classnames.element("featuredItem")}
            container
            item
            xs={6}
            spacing={1}
          >
            <Grid
              item
              xs={12}
              className={classnames.element("featuredItemContent")}
            >
              <BrowseCard data={featuredItem} variant="lg" withPrice={false} />
            </Grid>
          </Grid>
          <Grid
            className={classnames.element("subFeaturedItemsContainer")}
            container
            item
            spacing={1}
            xs={6}
          >
            {subFeaturedItems.map(item => (
              <Grid
                className={classnames.element("subFeaturedItem")}
                key={item.appid}
                item
                xs={6}
              >
                <BrowseCard data={item} variant="lg" withPrice={false} />
              </Grid>
            ))}
          </Grid>
        </Grid>
      ) : (
        <React.Fragment>
          <Grid
            className={classnames.element("featuredItemContainer")}
            container
            alignItems={"center"}
          >
            {items[currentPage - 1] && (
              <Grid
                className={classnames.element("featuredItem", {
                  fullWidth: true
                })}
                item
                xs={12}
              >
                <BrowseCard
                  data={items[currentPage - 1]}
                  variant="header"
                  withPrice={false}
                  maxSize={true}
                />
              </Grid>
            )}
          </Grid>
          <PaginationWidget
            className={classnames.element("paginationWidget")}
            currentPage={currentPage}
            setCurrentPage={setCurrentPage}
            totalPages={items.length}
          />
        </React.Fragment>
      )}
    </FeaturedWidget>
  );
};
MostPopularFeatured.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}MostPopularFeatured`
);
MostPopularFeatured.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    featuredItemContainer: PropTypes.string,
    featuredItem: PropTypes.string,
    featuredItemContent: PropTypes.string,
    featuredItemFullWidth: PropTypes.string,
    subFeaturedItemsContainer: PropTypes.string,
    subFeaturedItem: PropTypes.string,
    featuredItemContainer: PropTypes.string,
    featuredItem: PropTypes.string,
    paginationWidget: PropTypes.string
  }),
  // data for the cards to be rendered in the FeaturedWidget
  items: PropTypes.array,
  // material-ui width
  width: PropTypes.string
};
MostPopularFeatured.defaultProps = {
  classes: {}
};

// apply styles
export const StyledMostPopularFeatured = withWidth()(
  withStyles(styles)(MostPopularFeatured)
);

export default StyledMostPopularFeatured;
