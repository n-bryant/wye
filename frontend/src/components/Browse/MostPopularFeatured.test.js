import { MostPopularFeatured } from "./MostPopularFeatured";

describe("MostPopularFeatured", () => {
  const props = {
    classes: {
      featuredItemContainer: "featuredItemContainer",
      featuredItem: "featuredItem",
      featuredItemContent: "featuredItemContent",
      subFeaturedItemsContainer: "subFeaturedItemsContainer",
      subFeaturedItem: "subFeaturedItem",
      featuredItemContainer: "featuredItemContainer",
      featuredItem: "featuredItem",
      paginationWidget: "paginationWidget",
      featuredItemFullWidth: "featuredItemFullWidth",
      quickLinkContainer: "quickLinkContainer"
    },
    items: [
      {
        appid: "1"
      },
      {
        appid: "2"
      },
      {
        appid: "3"
      },
      {
        appid: "4"
      },
      {
        appid: "5"
      }
    ],
    width: "md"
  };

  it("should render successfully", () => {
    expect(
      toJson(shallow(<MostPopularFeatured {...props} />))
    ).toMatchSnapshot();
  });

  it("should render a full width card with pagination widget if the width is xs or sm", () => {
    const smallWidths = ["xs", "sm"];
    smallWidths.forEach(width => {
      let propsWithSmallWidth = {
        ...props,
        width
      };

      const wrapper = shallow(<MostPopularFeatured {...propsWithSmallWidth} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
