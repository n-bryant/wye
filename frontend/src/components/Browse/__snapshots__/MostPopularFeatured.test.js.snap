// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MostPopularFeatured should render a full width card with pagination widget if the width is xs or sm 1`] = `
<WithStyles(FeaturedWidget)
  className="wye-MostPopularFeatured"
  classes={Object {}}
  subTitle="Most popular by recent playtime"
  title="Trending"
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    className="wye-MostPopularFeatured__featuredItemContainer featuredItemContainer"
    container={true}
  >
    <WithStyles(ForwardRef(Grid))
      className="wye-MostPopularFeatured__featuredItem wye-MostPopularFeatured__featuredItem--fullWidth featuredItem featuredItemFullWidth"
      item={true}
      xs={12}
    >
      <HighlightTrailerQuery
        data={
          Object {
            "appid": "1",
          }
        }
        maxSize={true}
        variant="header"
        withPrice={false}
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(PaginationWidget)
    className="wye-MostPopularFeatured__paginationWidget paginationWidget"
    classes={Object {}}
    currentPage={1}
    setCurrentPage={[Function]}
    totalPages={5}
  />
  <Styled(MuiBox)
    className="wye-MostPopularFeatured__quickLinkContainer quickLinkContainer"
    my={4}
  >
    <WithStyles(QuickLink)
      classes={Object {}}
      label="Top 100"
      linkHref="/browse/top100"
    />
  </Styled(MuiBox)>
</WithStyles(FeaturedWidget)>
`;

exports[`MostPopularFeatured should render a full width card with pagination widget if the width is xs or sm 2`] = `
<WithStyles(FeaturedWidget)
  className="wye-MostPopularFeatured"
  classes={Object {}}
  subTitle="Most popular by recent playtime"
  title="Trending"
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    className="wye-MostPopularFeatured__featuredItemContainer featuredItemContainer"
    container={true}
  >
    <WithStyles(ForwardRef(Grid))
      className="wye-MostPopularFeatured__featuredItem wye-MostPopularFeatured__featuredItem--fullWidth featuredItem featuredItemFullWidth"
      item={true}
      xs={12}
    >
      <HighlightTrailerQuery
        data={
          Object {
            "appid": "1",
          }
        }
        maxSize={true}
        variant="header"
        withPrice={false}
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(PaginationWidget)
    className="wye-MostPopularFeatured__paginationWidget paginationWidget"
    classes={Object {}}
    currentPage={1}
    setCurrentPage={[Function]}
    totalPages={5}
  />
  <Styled(MuiBox)
    className="wye-MostPopularFeatured__quickLinkContainer quickLinkContainer"
    my={4}
  >
    <WithStyles(QuickLink)
      classes={Object {}}
      label="Top 100"
      linkHref="/browse/top100"
    />
  </Styled(MuiBox)>
</WithStyles(FeaturedWidget)>
`;

exports[`MostPopularFeatured should render successfully 1`] = `
<WithStyles(FeaturedWidget)
  className="wye-MostPopularFeatured"
  classes={Object {}}
  subTitle="Most popular by recent playtime"
  title="Trending"
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    className="wye-MostPopularFeatured__featuredItemContainer featuredItemContainer"
    container={true}
    justify="space-between"
  >
    <WithStyles(ForwardRef(Grid))
      className="wye-MostPopularFeatured__featuredItem featuredItem"
      container={true}
      item={true}
      spacing={1}
      xs={6}
    >
      <WithStyles(ForwardRef(Grid))
        className="wye-MostPopularFeatured__featuredItemContent featuredItemContent"
        item={true}
        xs={12}
      >
        <HighlightTrailerQuery
          data={
            Object {
              "appid": "1",
            }
          }
          variant="lg"
          withPrice={false}
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="wye-MostPopularFeatured__subFeaturedItemsContainer subFeaturedItemsContainer"
      container={true}
      item={true}
      spacing={1}
      xs={6}
    >
      <WithStyles(ForwardRef(Grid))
        className="wye-MostPopularFeatured__subFeaturedItem subFeaturedItem"
        item={true}
        key="2"
        xs={6}
      >
        <HighlightTrailerQuery
          data={
            Object {
              "appid": "2",
            }
          }
          variant="lg"
          withPrice={false}
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="wye-MostPopularFeatured__subFeaturedItem subFeaturedItem"
        item={true}
        key="3"
        xs={6}
      >
        <HighlightTrailerQuery
          data={
            Object {
              "appid": "3",
            }
          }
          variant="lg"
          withPrice={false}
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="wye-MostPopularFeatured__subFeaturedItem subFeaturedItem"
        item={true}
        key="4"
        xs={6}
      >
        <HighlightTrailerQuery
          data={
            Object {
              "appid": "4",
            }
          }
          variant="lg"
          withPrice={false}
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="wye-MostPopularFeatured__subFeaturedItem subFeaturedItem"
        item={true}
        key="5"
        xs={6}
      >
        <HighlightTrailerQuery
          data={
            Object {
              "appid": "5",
            }
          }
          variant="lg"
          withPrice={false}
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <Styled(MuiBox)
    className="wye-MostPopularFeatured__quickLinkContainer quickLinkContainer"
    my={4}
  >
    <WithStyles(QuickLink)
      classes={Object {}}
      label="Top 100"
      linkHref="/browse/top100"
    />
  </Styled(MuiBox)>
</WithStyles(FeaturedWidget)>
`;
